### PR TITLE
make `node:fs.ReadableStream` and `node:stream.NativeReadable` constructor functions

### PR DESCRIPTION
**Pull request by nektro**
Closes https://github.com/oven-sh/bun/issues/4417

### CODE REVIEW

File: src/js/node/fs.ts [hunk 1]
```diff
@@ -535,341 +535,309 @@ var defaultReadStreamOptions = {
...
-let kHandle = Symbol("kHandle");
-
 const blobToStreamWithOffset = $newZigFunction("blob.zig", "Blob.toStreamWithOffset", 1);
 
-var ReadStreamClass;
+function createReadStream(path, options) {
+  return new ReadStream(path, options);
+}
+
+const NativeReadable = Stream._getNativeReadableStreamPrototype(2, Stream.Readable);
```
Comment (Jarred-Sumner) [inline src/js/node/fs.ts:544]:
```suggestion
const NativeReadable = Stream._getNativeReadableStreamPrototype(2, Stream.Readable);
const NativeReadablePrototype = NativeReadable.prototype;
```

File: src/js/node/http.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@
 const EventEmitter = require("node:events");
 const { isTypedArray } = require("node:util/types");
 const { Duplex, Readable, Writable } = require("node:stream");
+const util = require("node:util");
```
Comment (Jarred-Sumner) [inline src/js/node/http.ts]:
```suggestion
```

File: src/js/node/http.ts [hunk 3]
```diff
@@ -476,6 +477,10 @@ Server.prototype.close = function (optionalCallback?) {
   this.emit("close");
 };
 
+Server.prototype[Symbol.asyncDispose] = async function () {
+  return util.promisify(this.close).$apply(this);
```
Comment (Jarred-Sumner) [inline src/js/node/http.ts]:
```suggestion
  const {promise, resolve, reject} = Promise.withResolvers();
  this.close((err, ...args) => {
     err ? reject(err) : resolve(...args);
  });
```

File: src/js/node/tty.ts [hunk 4]
```diff
@@ -4,29 +4,26 @@ const {
   getWindowSize: _getWindowSize,
 } = $cpp("ProcessBindingTTYWrap.cpp", "createBunTTYFunctions");
 
+const fs = require("node:fs");
```
Comment (Jarred-Sumner) [inline src/js/node/tty.ts]: lets keep the lazy initialization as before

File: src/js/node/tty.ts [hunk 5]
```diff
@@ -4,29 +4,26 @@ const {
...
-
-  return stream;
+  this.isRaw = false;
+  this.isTTY = true;
 }
 
 Object.defineProperty(ReadStream, "prototype", {
   get() {
-    const Prototype = Object.create(require("node:fs").ReadStream.prototype);
+    const Prototype = Object.create(fs.ReadStream.prototype);
```
Comment (Jarred-Sumner) [inline src/js/node/tty.ts]:
```suggestion
    const Prototype = Object.create(require("node:fs").ReadStream.prototype);
```

File: src/js/node/tty.ts [hunk 6]
```diff
@@ -142,7 +139,7 @@ function WriteStream(fd) {
   if (!(this instanceof WriteStream)) return new WriteStream(fd);
   if (fd >> 0 !== fd || fd < 0) throw new RangeError("fd must be a positive integer");
 
-  const stream = require("node:fs").WriteStream.$call(this, "", { fd });
+  const stream = fs.WriteStream.$call(this, "", { fd });
```
Comment (Jarred-Sumner) [inline src/js/node/tty.ts]: undo this please

File: src/js/node/tty.ts [hunk 7]
```diff
@@ -161,7 +158,7 @@ function WriteStream(fd) {
 
 Object.defineProperty(WriteStream, "prototype", {
   get() {
-    const Real = require("node:fs").WriteStream.prototype;
+    const Real = fs.WriteStream.prototype;
```
Comment (Jarred-Sumner) [inline src/js/node/tty.ts]: undo this please

File: test/js/node/http/node-http.test.ts [hunk 8]
```diff
@@ -2084,3 +2084,60 @@ it("ServerResponse ClientRequest field exposes agent getter", async () => {
     server.close();
   }
 });
+
+it("'st' package works", async () => {
```
Comment (Jarred-Sumner) [inline test/js/node/http/node-http.test.ts]: can we move this into a separate test file so it's clearer that its an integration test when it is flaky?

File: src/js/node/fs.ts [hunk 9]
```diff
@@ -535,341 +535,308 @@ var defaultReadStreamOptions = {
...
-        }
-      }
+  // At this point, n is the lesser of the length of the rest of the file to read or the highwatermark
+  // Which means n is the maximum number of bytes to read
+
+  // Basically if we don't know the file size yet, then check it
+  // Then if n is bigger than fileSize, set n to be fileSize
+  // This is a fast path to avoid allocating more than the file size for a small file (is this respected by native stream though)
+  if (this[mfileSize] === -1 && bytesRead === 0 && pos === undefined) {
+    var stat = fstatSync(fd);
```
Comment (Jarred-Sumner) [inline src/js/node/fs.ts:760]: we should remove this code later, we already call stat in Bun.file(path).stream

not in this PR though

File: src/js/node/http.ts [hunk 10]
```diff
@@ -476,6 +476,15 @@ Server.prototype.close = function (optionalCallback?) {
   this.emit("close");
 };
 
+Server.prototype[Symbol.asyncDispose] = async function () {
```
Comment (Jarred-Sumner) [inline src/js/node/http.ts]:
```suggestion
Server.prototype[Symbol.asyncDispose] = function () {
```

File: src/js/node/tty.ts [hunk 11]
```diff
@@ -8,20 +8,15 @@ const {
...
   if (fd >> 0 !== fd || fd < 0) throw new RangeError("fd must be a positive integer");
 
-  const stream = require("node:fs").ReadStream.$call(this, "", {
-    fd,
-  });
-  Object.setPrototypeOf(stream, ReadStream.prototype);
-
-  stream.isRaw = false;
-  stream.isTTY = true;
+  require("node:fs").ReadStream.$apply(this, ["", { fd }]);
```
Comment (Jarred-Sumner) [inline src/js/node/tty.ts:16]:
```suggestion
  require("node:fs").ReadStream.$apply(this, ["", { fd }]);
  Object.setPrototypeOf(stream, ReadStream.prototype);
```

**Comment by Jarred-Sumner** [comment]
build errors 
2 nitpicky comments

File: src/js/node/tty.ts (hunk 11 above)
Comment (nektro) [reply src/js/node/tty.ts:16]: `stream` doesnt exist anymore

**Pull request merged**